File: SAN/Auth/AuthService.swift
import Foundation

// MARK: - Пользователь

struct SANUser: Codable, Equatable {
    let id: String
    var name: String
    var email: String?
    var provider: AuthProvider
}

enum AuthProvider: String, Codable {
    case apple, google, email, guest
}

// MARK: - Ошибки

enum AuthError: LocalizedError {
    case cancelled
    case invalidCredentials
    case notConfigured(String)
    case unknown(String)

    var errorDescription: String? {
        switch self {
        case .cancelled: return "Вход отменён"
        case .invalidCredentials: return "Неверная почта или пароль"
        case .notConfigured(let p): return "\(p) ещё не подключён. Нужен Firebase-проект и ключи."
        case .unknown(let m): return m
        }
    }
}

// MARK: - Протокол сервиса авторизации
// Mock-реализация работает сейчас. FirebaseAuthService подключается позже
// без изменения UI — он реализует тот же протокол.

protocol AuthService {
    func currentUser() -> SANUser?
    /// Firebase ID-токен текущего пользователя (для авторизации вызовов Cloud Functions).
    func idToken() async -> String?
    func signInWithEmail(_ email: String, password: String) async throws -> SANUser
    func registerWithEmail(name: String, email: String, password: String) async throws -> SANUser
    func signInWithGoogle() async throws -> SANUser
    /// Результат нативного Sign in with Apple. idTokenString + rawNonce нужны
    /// для обмена на Firebase-credential; mock использует только id/имя/email.
    func signInWithApple(_ credential: AppleCredential) async throws -> SANUser
    /// Анонимный вход (Firebase signInAnonymously)
    func continueAsGuest() async throws -> SANUser
    func signOut()
}

/// Данные, полученные от Sign in with Apple.
struct AppleCredential {
    let userID: String
    let idTokenString: String?
    let rawNonce: String?
    let name: String?
    let email: String?
}

File: android/app/src/main/java/kg/ayant/app/data/AuthService.kt
package kg.ayant.app.data

import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.tasks.await
import kg.ayant.app.ui.vm.AuthProvider
import kg.ayant.app.ui.vm.AyantUser
import java.util.UUID

/**
 * Auth abstraction. MockAuthService works offline; FirebaseAuthService uses the
 * shared san-25d32 Auth. Mirrors AuthService.swift (Apple omitted — iOS-only).
 */
interface AuthService {
    fun currentUser(): AyantUser?
    /** Firebase ID token for authorizing Cloud Function calls (scanCoupon). */
    suspend fun idToken(): String?
    suspend fun signInEmail(email: String, password: String): AyantUser
    suspend fun registerEmail(name: String, email: String, password: String): AyantUser
    suspend fun signInGoogle(): AyantUser
    suspend fun continueAsGuest(): AyantUser
    fun signOut()
}

/** Offline mock — accepts any input, derives a stable id from email. */
class MockAuthService : AuthService {
    override fun currentUser(): AyantUser? = null
    override suspend fun idToken(): String? = "mock-token"
    override suspend fun signInEmail(email: String, password: String): AyantUser {
        require(email.isNotBlank() && password.isNotBlank()) { "Введите почту и пароль" }
        return AyantUser(stableId(email), nameFromEmail(email), email, AuthProvider.EMAIL)
    }
    override suspend fun registerEmail(name: String, email: String, password: String): AyantUser {
        require(name.isNotBlank() && email.isNotBlank() && password.isNotBlank()) { "Заполните все поля" }
        return AyantUser(stableId(email), name, email, AuthProvider.EMAIL)
    }
    override suspend fun signInGoogle(): AyantUser =
        AyantUser(UUID.randomUUID().toString(), "Пользователь Google", null, AuthProvider.GOOGLE)
    override suspend fun continueAsGuest(): AyantUser =
        AyantUser("guest_${UUID.randomUUID().toString().take(8)}", "Гость", null, AuthProvider.GUEST)
    override fun signOut() {}
    private fun nameFromEmail(e: String) = e.substringBefore("@").replaceFirstChar { it.uppercase() }
    private fun stableId(e: String) = "u_" + e.lowercase().hashCode().toUInt().toString(16)
}

/** Firebase Auth (email + anonymous). Google needs Credential Manager — falls back to mock. */
class FirebaseAuthService : AuthService {
    private val auth = FirebaseAuth.getInstance()

    override fun currentUser(): AyantUser? = auth.currentUser?.let {
        AyantUser(it.uid, it.displayName ?: it.email?.substringBefore("@") ?: "Вы", it.email,
            if (it.isAnonymous) AuthProvider.GUEST else AuthProvider.EMAIL)
    }

    override suspend fun idToken(): String? =
        runCatching { auth.currentUser?.getIdToken(false)?.await()?.token }.getOrNull()

    override suspend fun signInEmail(email: String, password: String): AyantUser {
        val res = auth.signInWithEmailAndPassword(email, password).await()
        val u = res.user!!
        return AyantUser(u.uid, u.displayName ?: nameFromEmail(email), u.email, AuthProvider.EMAIL)
    }

    override suspend fun registerEmail(name: String, email: String, password: String): AyantUser {
        val res = auth.createUserWithEmailAndPassword(email, password).await()
        val u = res.user!!
        return AyantUser(u.uid, name, u.email, AuthProvider.EMAIL)
    }

    override suspend fun signInGoogle(): AyantUser =
        // Google Sign-In (Credential Manager) is a later pass; use anonymous for now.
        continueAsGuest()

    override suspend fun continueAsGuest(): AyantUser {
        val res = auth.signInAnonymously().await()
        return AyantUser(res.user!!.uid, "Гость", null, AuthProvider.GUEST)
    }

    override fun signOut() { auth.signOut() }

    private fun nameFromEmail(e: String) = e.substringBefore("@").replaceFirstChar { it.uppercase() }
}
